File: iloaders/replace_old_adsr.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2026 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static void replace_old_adsr (uint8_t patterns)
{
  tFIXED_SONGDATA *song = &songdata;
  uint8_t order = 0;
  uint8_t patts_count = 0;
  uint8_t patts[255], patt;

  memset (adsr_carrier, false, sizeof (adsr_carrier));

  do
  {
    patt = song->pattern_order[order];

    if (patt >= 0x80) order++;
    else
    {
      uint8_t patt_break = BYTE_NULL;

      for (uint8_t i = 0; i < 0x40; i++)
        for (uint8_t j = 0; j < 9; j++)
        {
          tCHUNK chunk, chunk2;

          get_chunk (patt, i, j + 1, &chunk);
          chunk2 = chunk;

          if (   (chunk.effect_def == ef_PositionJump)
              || (chunk.effect_def == ef_PatternBreak))
            patt_break = i;

          if ((chunk.effect_def == 0xFF) || (chunk.effect_def == ef_Extended))
          {
            if (chunk.effect_def == 0xFF)
            {
              chunk2.effect_def = 0;
              chunk2.effect = 0;

              if (i <= patt_break)
                switch (chunk.effect)
                {
                  case 0: adsr_carrier[j] = true; break;
                  case 1: adsr_carrier[j] = false; break;
                  default: break;
                }
            }

            if (chunk.effect_def == ef_Extended)
              switch (chunk.effect >> 4)
              {
                case ef_ex_SetAttckRateM:
                case ef_ex_SetAttckRateC:
                  chunk2.effect = ((adsr_carrier[j] ? ef_ex_SetAttckRateC : ef_ex_SetAttckRateM) << 4) + (chunk.effect & 0x0F);
                  break;

                case ef_ex_SetDecayRateM:
                case ef_ex_SetDecayRateC:
                  chunk2.effect = ((adsr_carrier[j] ? ef_ex_SetDecayRateC : ef_ex_SetDecayRateM) << 4) + (chunk.effect & 0x0F);
                  break;

                case ef_ex_SetSustnLevelM:
                case ef_ex_SetSustnLevelC:
                  chunk2.effect = ((adsr_carrier[j] ? ef_ex_SetSustnLevelC : ef_ex_SetSustnLevelM) << 4) + (chunk.effect & 0x0F);
                  break;

                case ef_ex_SetRelRateM:
                case ef_ex_SetRelRateC:
                  chunk2.effect = ((adsr_carrier[j] ? ef_ex_SetRelRateC : ef_ex_SetRelRateM) << 4) + (chunk.effect & 0x0F);
                  break;

                default: break;
              }

            if ((patts_count == 0) || (memchr (patts, patt, patts_count) == NULL))
              if ((chunk.effect_def != chunk2.effect_def) || (chunk.effect != chunk2.effect))
                put_chunk (patt, i, j + 1, &chunk2);
          }
        }

      order++;
      if (patts_count < sizeof (patts)) patts[patts_count++] = patt;
    }
  } while ((patt < patterns) || (order >= 0x80));
}

File: formats/pas/replace_old_adsr.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2026 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure replace_old_adsr(patterns: Byte);

var
  chunk,chunk2: tCHUNK;
  temp2,temp3: Byte;
  patt_break: Byte;
  order,patt: Byte;
  patts: String;

begin
  patts := '';
  FillChar(adsr_carrier,SizeOf(adsr_carrier),0);

  order := 0; patt := BYTE_NULL;
  Repeat
    If (songdata.pattern_order[order] >= $80) then Inc(order)
    else
      begin
        patt := songdata.pattern_order[order];
        patt_break := BYTE_NULL;
        For temp2 := 0 to $3f do
          For temp3 := 1 to 9 do
            begin
              get_chunk(patt,temp2,temp3,chunk);
              chunk2 := chunk;

              If (chunk.effect_def in [ef_PositionJump,ef_PatternBreak]) then
                patt_break := temp2;

              If (chunk.effect_def in [$ff,ef_Extended]) then
                begin
                  If (chunk.effect_def = $ff) then
                    begin
                      chunk2.effect_def := 0;
                      chunk2.effect := 0;

                      If (temp2 <= patt_break) then
                        Case chunk.effect of
                          0: adsr_carrier[temp3] := TRUE;
                          1: adsr_carrier[temp3] := FALSE;
                        end;
                    end;

                  If (chunk.effect_def = ef_Extended) then
                    Case chunk.effect DIV 16 of
                      ef_ex_SetAttckRateM,
                      ef_ex_SetAttckRateC:
                        If adsr_carrier[temp3] then
                          chunk2.effect := ef_ex_SetAttckRateC*16+chunk.effect MOD 16
                        else chunk2.effect := ef_ex_SetAttckRateM*16+chunk.effect MOD 16;

                      ef_ex_SetDecayRateM,
                      ef_ex_SetDecayRateC:
                        If adsr_carrier[temp3] then
                          chunk2.effect := ef_ex_SetDecayRateC*16+chunk.effect MOD 16
                        else chunk2.effect := ef_ex_SetDecayRateM*16+chunk.effect MOD 16;

                      ef_ex_SetSustnLevelM,
                      ef_ex_SetSustnLevelC:
                        If adsr_carrier[temp3] then
                          chunk2.effect := ef_ex_SetSustnLevelC*16+chunk.effect MOD 16
                        else chunk2.effect := ef_ex_SetSustnLevelM*16+chunk.effect MOD 16;

                      ef_ex_SetRelRateM,
                      ef_ex_SetRelRateC:
                        If adsr_carrier[temp3] then
                          chunk2.effect := ef_ex_SetRelRateC*16+chunk.effect MOD 16
                        else chunk2.effect := ef_ex_SetRelRateM*16+chunk.effect MOD 16;
                    end;

                  If (Pos(CHR(songdata.pattern_order[order]),patts) = 0) then
                    If (chunk.effect_def <> chunk2.effect_def) or
                       (chunk.effect <> chunk2.effect) then
                      put_chunk(patt,temp2,temp3,chunk2);
                end;
            end;
        Inc(order);
        patts := patts+CHR(patt);
      end;
  until (patt >= patterns) or (order > $7f);
end;
